tenantrate: switch to a single token bucket

RaduBerinde opened the pull request:
Tenant rate limiting currently uses four separate token buckets, each
with its own rate and burst limits. In this commit we switch to a
single shared token bucket (see #55114 for motivation).

We use a "cost model" to map read and write requests to "KV Compute
Units". Requests have a base cost plus a per-byte cost. The details
are documented in settings.go. The values were chosen based on
experiments ran by Nathan:
https://docs.google.com/spreadsheets/d/1PPlIcKnusOqWtBoOZVd9xBEMPe5Ss1FgrJlYFgpQZaM/edit#gid=735409177

The rate was chosen so that it maps to 20% of 1 KV vCPU. This keeps
the rate limit on small requests roughly the same as before
(especially on mixed workloads).

The largest departure from the previous limits is that we allow much
more read bytes (the per-byte cost of reads is small in the cost
model). If we were to keep close to the previous limits, the value of
kv.tenant_rate_limiter.read_cost_per_megabyte would be 200 instead of
10.  Perhaps we want to be more conservative here and make this
value somewhere in-between?

Fixes #55114.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62783)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
Now that it's back to being one dimension, I can see an argument to just bend the quotapool.RateLimiter to your will. I have had a hard time striking the right abstraction level for that library. It's currently got these very generic hooks plus simple reference implementations which are pretty powerful on their own. It probably ought to be split out into a few package.

A problem with that is you'd need to either early-bind the mapping to resources or add an `AcquireFunc` equivalent to the `quotapool.IntPool` to read some configuration and you'd need to add support for the non-blocking subtraction. I feel like there's probably a way to work it such that your library here gets control of its Request and the settings and the abstraction there owns the rate and burst. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MX0KnTJBznLC7qmComi:b-77jhpj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @nvanbenschoten, @RaduBerinde, and @tbg)

___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MX0LLzoCa2GrJ4ulNs9:-MX0LLzoCa2GrJ4ulNsA:b-8847ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/776935342306d5f4cb997539d72298a7b56443eb/pkg/kv/kvserver/tenantrate/limiter.go#L99)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	// TODO(radu): find a way to omit these atomic operations in the case when we
> 	// don't have to wait.
> 	rl.metrics.currentBlocked.Inc(1)
> 	defer rl.metrics.currentBlocked.Dec(1)
> ```
</details>


Feel free to add the hooks to the quotapool library. We have an `OnAcquisition` hook that gets called regardless of whether or not you hit the fast path. It seems totally reasonable to add hooks for when you don't get the fast path and then when you finish waiting. The setup would go in config.go. The calls would be here:

https://github.com/cockroachdb/cockroach/blob/aec84caa37828f4d27c5707250150b55c47cdb66/pkg/util/quotapool/quotapool.go#L234
___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MX0NEQaERpPEANhUXZl:-MX0NEQaERpPEANhUXZm:b-bzu6b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/776935342306d5f4cb997539d72298a7b56443eb/pkg/kv/kvserver/tenantrate/limiter.go#L206)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> // TODO(ajwerner): It seems possible that when adding or reducing the burst
> // values that we might want to remove those values from the token bucket.
> // It's not obvious that we want to add tokens when increasing the burst as
> // that might lead to a big spike in load immediately upon increasing this
> // limit.
> ```
</details>

The code very much does this in the quotapool rate limter and I think it's worthwhile: https://github.com/cockroachdb/cockroach/blob/aec84caa37828f4d27c5707250150b55c47cdb66/pkg/util/quotapool/int_rate.go#L122-L127


<!-- Sent from Reviewable.io -->

tbg commented:
Can we take a step back and go through what the plan here is? When I filed that issue I still had the distributed token bucket and "protecting KV" mixed up in my head. The KV-side rate limiter is strictly for "protecting KV", right? As such, is there any reason to bake compute units into it? Won't this per-tenant limiter go away and be replaced by a per-node limiter and would we then limit on "request units" on a node level which doesn't necessarily seem that useful? Or are we going to keep the per-tenant limiters around on top of that and simply set much higher limits than today (i.e. allow a tenant to exhaust, say, 30% of the node assuming it has capacity)?

I don't think there is a single place that really spells out what the plan is (right?), which worries me.

andy-kimball reviewed (commented):
Actually, there's a 3rd concern: limit each tenant to some % of the KV machine. We never want to allow any single tenant to monopolize KV resources, _even if there are no other tenants running on that machine_. The reason is that this will lead to severe "noisy neighbor" syndrome, where tenants will see amazing performance when they happen to be the only ones running on a KV node, followed by sharp dips in performance if just 1 or 2 other tenants begins running. If instead we limit each tenant to a maximum of say 10% of a KV node (and I expect these nodes to be big), then adding more than 10 tenants leads to a much more incremental drop in performance.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MX2S4gd8V7eXmQg5qT-:bz6q2b4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @RaduBerinde, and @tbg)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
We still want to limit the amount of resources used by any one free tier user, regardless of whether the KV nodes are being overloaded or not. We will add a separate mechanism for limiting per-KV-node load, and this tenantrate mechanism will evolve to perform distributed rate-limiting (where the limit will apply per cluster, not per KV node).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MX2x13W8J-J-JqFZdoV:bokebwi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @RaduBerinde, and @tbg)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> Now that it's back to being one dimension, I can see an argument to just bend the quotapool.RateLimiter to your will [...]

Thanks, I will take a look - this is my first attempt at working with this code. Some things did feel awkward to me, especially the plumbing around `readBytesResource` when all we really need is to modify the tenantBucket under the quotapool's lock (but I didn't have any epiphany on how to make it simpler).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MX3LmP58gRE7Sj2kzIZ:by2cp1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @RaduBerinde, and @tbg)



<!-- Sent from Reviewable.io -->

tbg commented:
> We still want to limit the amount of resources used by any one free tier user, regardless of whether the KV nodes are being overloaded or not. We will add a separate mechanism for limiting per-KV-node load, and this tenantrate mechanism will evolve to perform distributed rate-limiting (where the limit will apply per cluster, not per KV node).

Thanks for explaining Andy and Radu. And the argument for using resource units is what? The motivation in https://github.com/cockroachdb/cockroach/issues/55114 stemmed from the fact that I thought these would be surfaced externally, but isn't the per-tenant limiter at KV internal? Does it make sense to compile these constants we've chosen into CRDB, where we can't change them on a per-tenant basis?

RaduBerinde commented:
We want the limits to be proportional to the actual resource usage (and thus how much we pay for the KV nodes), as much as possible. Under the current model, a mixed workload can get twice as many resources than a read-heavy or a write-heavy workload, which doesn't make sense. It is also a big advantage to use a single token bucket because the next step will be implementing it in a distributed fashion and I don't want 4x the complication and overhead.

Per-tenant tunability is perpendicular, we don't have that today but we can easily add it with either approach.

I would also turn it around on you and ask why you think the four token buckets are better?

RaduBerinde reviewed (commented):
Made some updates. Please ignore the first commit, it's from a separate PR that I've been trying to merge all day.

It should be interesting to look at the diff in `estimate_iops`, it will show how various workload profiles would perform now vs before.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MX9Mlm22CCAeSfonz8h:btjckol)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MX0NEQaERpPEANhUXZl:-MX8q2uX3VyaQYWF5_qG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/776935342306d5f4cb997539d72298a7b56443eb/pkg/kv/kvserver/tenantrate/limiter.go#L206)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> // TODO(ajwerner): It seems possible that when adding or reducing the burst
> // values that we might want to remove those values from the token bucket.
> // It's not obvious that we want to add tokens when increasing the burst as
> // that might lead to a big spike in load immediately upon increasing this
> // limit.
> ```
The code very much does this in the quotapool rate limter and I think it's worthwhile: https://github.com/cockroachdb/cockroach/blob/aec84caa37828f4d27c5707250150b55c47cdb66/pkg/util/quotapool/int_rate.go#L122-L127
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
One thing this discussion highlights: do we need the allocator to try and spread out ranges for a given tenant across as many nodes as possible? Otherwise, if a tenant's ranges all happen to be on the same node, they will hit their per-tenant throttling limit earlier than if the pages are more spread out. And the activity from that one tenant may not be enough to trigger the allocator to move pages around either, so it will just stay in a permanent state of throttling.

RaduBerinde commented:
That will not matter once we have the distributed rate limiting, correct? It sounds like a good idea to me in general though.. spreading things as much as possible reduces the magnitude of the noisy neighbor problem on any one node.

andy-kimball commented:
I think it will still be a problem even with distributed rate limiting. The per-tenant KV rate limit would throttle a given tenant's pods on a particular machine, even if the distributed rate limiter has plenty of available capacity for that tenant.

RaduBerinde commented:
Hm, in my mind we would implement only distributed per-tenant rate limiting, where we control the total rate across all KV nodes. There wouldn't also be a per-node per-tenant rate limit.

We would have a separate per-node aggregate tenant limit to avoid overload (and for the purpose of avoiding overload, I agree that spreading out would help).

andy-kimball commented:
I see. My worry with that is that then tenants would see a different problem. If they happened to be the only tenant running on a KV node, they'd get fabulous throughput. Then if even one other tenant starts running on the same KV node, the 1st tenant would see their throughput cut in half. So there could be fairly severe "noisy neighbor" effects. I was assuming the per-node limit would be per-tenant.

RaduBerinde commented:
A tenant would never do more than the distributed per-tenant rate limit, which I assume won't be too fabulous (probably a very small multiple of the current per-node limit, right?).

The per-node aggregate limit would be much higher than what a few tenants would be able to push through (the goal would be for this limit to rarely come into play - we would try to provision the cluster so that KV nodes don't routinely get overloaded).

andy-kimball commented:
I'm thinking about paid tenants that are not rate limited, because they're paying for whatever they use. They want to get as much performance as possible, but with a reasonable amount of predictability.

RaduBerinde commented:
Ah, I see. I suspect we may still want some limiting there too (at a much higher rate). But in any case, noisy paid neighbors will be problematic.

RaduBerinde committed:
tenantrate: switch to a single token bucket

Tenant rate limiting currently uses four separate token buckets, each
with its own rate and burst limits. In this commit we switch to a
single shared token bucket (see #55114 for motivation).

We use a "cost model" to map read and write requests to "KV Compute
Units". Requests have a base cost plus a per-byte cost. The details
are documented in settings.go. The values were chosen based on
experiments ran by Nathan:
https://docs.google.com/spreadsheets/d/1PPlIcKnusOqWtBoOZVd9xBEMPe5Ss1FgrJlYFgpQZaM/edit#gid=735409177

The rate was chosen so that it maps to 20% of 1 KV vCPU. This keeps
the rate limit on small requests roughly the same as before
(especially on mixed workloads).

The largest departure from the previous limits is that we allow much
more read bytes (the per-byte cost of reads is small in the cost
model). If we were to keep close to the previous limits, the value of
kv.tenant_rate_limiter.read_cost_per_megabyte would be 200 instead of
10.  Perhaps we want to be more conservative here and make this
value somewhere in-between?

Fixes #55114.

Release note: None

sumeerbhola reviewed (commented):
It's not completely clear to me how the code artifacts here will extend into a distributed rate limiting setting. Do you expect to reuse `limiter`, `tokenBucket` etc. in a distributed setting by having a local "controller" fill the tokens based on what it has pre-allocated from a global coordinator?


> If instead we limit each tenant to a maximum of say 10% of a KV node ...
> do we need the allocator to try and spread out ranges for a given tenant across as many nodes as possible? Otherwise, if a tenant's ranges all happen to be on the same node, they will hit their per-tenant throttling limit earlier than if the pages are more spread out. 

Regarding the noisy neighbor and per-tenant per-node limits etc. discussion above, there are some interesting points being made. I've typically not seen systems actively spreading a tenant's ranges across more nodes even if there is no need to based on current load -- one reason being that isolation failures do happen (and even worse, sometimes there are queries of death). Actively, spreading each tenant around increases the blast radius. The corollary being there aren't any limits on what a tenant can use of the resources in a node (or maybe something like 75%), and if another tenant wakes up and starts to consume more resources, there is a performance decline until rebalancing kicks in. It would be interesting to think through this and make improvements once we can observe real serverless clusters with significant tenant activity -- do our current serverless cluster(s) have enough utilization to make it worthwhile to look at them?

> Reviewed 2 of 10 files at r1, 4 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXcHEEwAH6QjfV8IsKO:b-g86gwr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, @RaduBerinde, and @tbg)

___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcDMcq7XB2QdKgl1l4:-MXcDMcq7XB2QdKgl1l5:b-wb2u12) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L136)):*
> ```Go
> 
> // updateConfig is used by the factory to inform the limiter of a new
> // configuration.
> ```

is this a new configuration, or a delta change to the configuration? It looks like a delta, based on the `Add` in the implementation.
___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcE_dC2fO4d0E0etmA:-MXcE_dC2fO4d0E0etmB:bh4b995) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L225)):*
> ```Go
> 	}
> }
> ```

This comment is not specific to this PR, but there seem to be many interfaces, structs and layers of abstraction in the rate limiting code, which makes it hard for me to wrap my head around the code. There is even `interface{}` in the method below, accompanied with a type switch.
Is there a high-level summary describing and motivating the need for all these code abstractions, for what to me naively boils down to what looks like a straightforward token bucket mechanism.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> It's not completely clear to me how the code artifacts here will extend into a distributed rate limiting setting. Do you expect to reuse limiter, tokenBucket etc. in a distributed setting by having a local "controller" fill the tokens based on what it has pre-allocated from a global coordinator?

I have some ideas that I hope to put in an RFC soon. I haven't yet thought far enough to know which parts of the code I'll reuse.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXcKV3F6rX6F0aK5xYA:bjkrr2x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, @RaduBerinde, @sumeerbhola, and @tbg)

___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcDMcq7XB2QdKgl1l4:-MXcKzow52uaKuJ5DATi:b9vb62e) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L136)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this a new configuration, or a delta change to the configuration? It looks like a delta, based on the `Add` in the implementation.
</blockquote></details>

It's a pretty non-intuitive API, `Add` just means that we will get a call to `Merge` with the `config` argument.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I would like to close on this soon.

My main hesitation is with the significantly higher read bytes allowance. In the spreadsheet @andy-kimball points out that the numbers may be skewed if the test dataset fits in memory. Should we set a more conservative value (in-between the current value and the one suggested by the experiments)?

ajwerner reviewed (commented):
I'm fine with the logic here. I worry the read unit is too high.

From a higher level, I'm worried we're missing a library at the right level of abstraction. The logic here is so similar to the `quotapool.RateLimiter`. The main difference is just where the synchronization happens on the configuration and the ability to suck quota out after the reads. It feels like if you just added a simple method to deal with the reads and then you did the conversion before queuing rather than after, then you could avoid all of this mucking with the bad `quotapool.QuotaPool` API. 

The only reason I can think to not do the above suggestion is because of an eye towards future changes, but I don't currently see any clear vision in that direction.

> Reviewed 2 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXeo0dCCCNwzA8Rq3Ro:b-5ltned)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, @RaduBerinde, @sumeerbhola, and @tbg)

___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcDMcq7XB2QdKgl1l4:-MXegVPKABM-fLAcT7DQ:bj2ue7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L136)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's a pretty non-intuitive API, `Add` just means that we will get a call to `Merge` with the `config` argument.
</blockquote></details>

I really struggled with how to frame this API. I agree it's extremely non-intuitive. A wanted to make a generic data structure that clients could leverage but really struggled to make something good. `Add` is a low-level primitive to muck with the state (`Resource`) that the `QuotaPool` is mediating access to on behalf of `Request`s. In the first pass of all of this is was just about providing an api to put resource back into the pool. In that world it was not so bad. Over time it got worse. I feel like the abstraction being offerred here is high leverage but I don't quite know how to package up the concepts properly. 
___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXenErS2dC71MwQUslm:-MXenErT1tlrp9UsAs8q:b-7daxv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L203)):*
> ```Go
> 	}
> 	delta := needed - tb.tokens
> 	tryAgainAfter = time.Duration((delta * float64(time.Second)) / tb.config.Rate)
> ```

There are cases where the rates are really high where having a fallback can matter. May as well productionize this code while we're here. Could see adding a test case. 

https://github.com/cockroachdb/cockroach/blob/c8331c0a9519d30bbb06295156513b0d1352c3ba/pkg/util/quotapool/int_rate.go#L193-L200
___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcE_dC2fO4d0E0etmA:-MXeieLTC66e5KSEGcmo:b5zzw3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L225)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This comment is not specific to this PR, but there seem to be many interfaces, structs and layers of abstraction in the rate limiting code, which makes it hard for me to wrap my head around the code. There is even `interface{}` in the method below, accompanied with a type switch.
Is there a high-level summary describing and motivating the need for all these code abstractions, for what to me naively boils down to what looks like a straightforward token bucket mechanism.
</blockquote></details>

I can take this. 

The basic idea is that a `quotapool.QuotaPool` should be used to implement some form of queue+limit logic. It is not intended for direct consumption. The `QuotaPool` protects some `Resource` which is handed out to `Request`s. The `Request` implementation gets handed the `Resource` and can modify it and report whether it is satisfied. In order to add rate-limiting functionality, the request can also indicate when it should get notified to try again. The other way the `Resource` can be modified is with the `Add()` method which passes an arbitrary object to the `Resource` instance under the `QuotaPool`'s mutex, and decides whether now the head of the queue should be notified to try again. It's all somewhat abstract.  

This code sort of grew, as code does. This code is used to implement both an integer-resource quotapool as well as a token bucket rate-limiter. Any thoughts on API improvements would be appreciated. The API surface area of `quotapool.IntPool` and `quotepool.RateLimiter` are both sane but the lower-level `quotapool.QuotaPool` is an opaque and non-intuitive. Also, the interfaces push towards a sync.Pool to avoid allocation which is a bummer but seemingly unavoidable in go when one wants both abstraction and performance. I have hopes that generics may make expressing some of these concepts better but I don't have a concrete vision. 

The entry point for the code is likely https://github.com/cockroachdb/cockroach/blob/c8331c0a9519d30bbb06295156513b0d1352c3ba/pkg/util/quotapool/quotapool.go#L32-L69
___
*[pkg/kv/kvserver/tenantrate/settings.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXelyfXBMU0ZCqh-biG:-MXelyfXBMU0ZCqh-biH:bst628x) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/settings.go#L39)):*
> ```Go
> 	Burst float64
> 
> 	// ReadRequestUnits is the baseline cost of a read, in KV Compute Units.
> ```

This comment could be more informative. The choice of value here tends to not be all that important but the fact that there is a value is very important. The explanation of the "courtesy" unit deserves to be somewhere and this feels like the spot. 
___
*[pkg/kv/kvserver/tenantrate/settings.go, line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXemi0YDhmfY0gN6g3D:-MXemi0YDhmfY0gN6g3E:b-fx4m0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/settings.go#L78)):*
> ```Go
> 		"kv.tenant_rate_limiter.read_request_cost",
> 		"base cost of a read request in KV Compute Units",
> 		0.7,
> ```

This feels high. 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> The logic here is so similar to the quotapool.RateLimiter.

I didn't notice earlier that `int_rate.go` had its own token bucket implementation. It would indeed be nice if we could reduce duplication.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXgWKsfAjRdi2q6_lVh:b-mdnw5c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, and @RaduBerinde)

___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcDMcq7XB2QdKgl1l4:-MXgXi2M4biVu1rbHhu3:b-mm38il) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L136)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I really struggled with how to frame this API. I agree it's extremely non-intuitive. A wanted to make a generic data structure that clients could leverage but really struggled to make something good. `Add` is a low-level primitive to muck with the state (`Resource`) that the `QuotaPool` is mediating access to on behalf of `Request`s. In the first pass of all of this is was just about providing an api to put resource back into the pool. In that world it was not so bad. Over time it got worse. I feel like the abstraction being offerred here is high leverage but I don't quite know how to package up the concepts properly. 
</blockquote></details>

I think this is what tripped me up when trying to understand these abstractions, which your other explanation has helped with. 
The abstraction is not really for abstracting the details of `Request` and `Resource` from each other since they know the specifics of the other. But their interaction is mediated by the `QuotaPool`, which is handling synchronization and queueing, which causes the type assertions (like `waitRequest` asserts that it is interacting with `tokenBuckets`).
The commentary around `Resource` is also quite confusing e.g. "It is any quantity that can be subdivided and combined", which is not true for the token buckets. There is also "After val is passed to Merge, the QuotaPool will never use that Resource again. This behavior allows clients to pool instances of Resources by creating Resource during Acquisition and destroying them in Merge." This doesn't seem right, though maybe it refers to the `val` parameter for `intAlloc` which is also an `intAlloc`.

I can see why generics would improve the situation, but the `Merge` method would still be clunky -- it is trying to do many things: config changes, unilateral resource acquisition, returning of resources for non-token-bucket. Perhaps separating them into different methods would aid understanding, though I'm not so sure that is needed, especially given your other comment about `QuotaPool` being for implementation sharing and not for "direct consumption".

I don't really have any radical ideas for restructuring, but code comment improvements would probably help the readability.
___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcE_dC2fO4d0E0etmA:-MXgb84k1nCwHIXQ3HMT:b-9xrplm) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L225)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I can take this. 

The basic idea is that a `quotapool.QuotaPool` should be used to implement some form of queue+limit logic. It is not intended for direct consumption. The `QuotaPool` protects some `Resource` which is handed out to `Request`s. The `Request` implementation gets handed the `Resource` and can modify it and report whether it is satisfied. In order to add rate-limiting functionality, the request can also indicate when it should get notified to try again. The other way the `Resource` can be modified is with the `Add()` method which passes an arbitrary object to the `Resource` instance under the `QuotaPool`'s mutex, and decides whether now the head of the queue should be notified to try again. It's all somewhat abstract.  

This code sort of grew, as code does. This code is used to implement both an integer-resource quotapool as well as a token bucket rate-limiter. Any thoughts on API improvements would be appreciated. The API surface area of `quotapool.IntPool` and `quotepool.RateLimiter` are both sane but the lower-level `quotapool.QuotaPool` is an opaque and non-intuitive. Also, the interfaces push towards a sync.Pool to avoid allocation which is a bummer but seemingly unavoidable in go when one wants both abstraction and performance. I have hopes that generics may make expressing some of these concepts better but I don't have a concrete vision. 

The entry point for the code is likely https://github.com/cockroachdb/cockroach/blob/c8331c0a9519d30bbb06295156513b0d1352c3ba/pkg/util/quotapool/quotapool.go#L32-L69

</blockquote></details>

Thanks, this is very helpful.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I will try to modify RateLimit. I was trying to limit the blast area of my change, given that we want to backport this. I will need to modify the burst to a float64 (not sure why it is int when the rate is float?) and add some more methods. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXgiero-2hz-3F-2l8e:bpfur1b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, and @RaduBerinde)

___
*[pkg/kv/kvserver/tenantrate/settings.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXelyfXBMU0ZCqh-biG:-MXgi8tF-6a_I68c0nGB:b-c1t3t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/settings.go#L39)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This comment could be more informative. The choice of value here tends to not be all that important but the fact that there is a value is very important. The explanation of the "courtesy" unit deserves to be somewhere and this feels like the spot. 
</blockquote></details>

The units are all explained in the comment above, what is it missing?
___
*[pkg/kv/kvserver/tenantrate/settings.go, line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXemi0YDhmfY0gN6g3D:-MXgiMVP9A7_75pPrVOJ:b-15mape) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/settings.go#L78)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This feels high. 
</blockquote></details>

Can you expand? What would you suggest and why?

If I was to stick close to the previous code, I would have it at 1.0. I lowered to 0.7 which is the value derived from experiments (column B6 times 1000 https://docs.google.com/spreadsheets/d/1PPlIcKnusOqWtBoOZVd9xBEMPe5Ss1FgrJlYFgpQZaM/edit#gid=735409177).


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Ack on the rate cost of a read request. I wasn't thinking about it properly. I was caught up thinking that most of the cost we wanted to calculate was on the data read than on the request itself and that this quantity was closer to the old "courtesy" byte concept that was here previously.

The only other user of the rate limiter is pretty straightforward. The units are integers because this code replaced some usages of https://pkg.go.dev/golang.org/x/time/rate which only allows integer quantities. I don't think floating point is going to be a problem. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXgrAn49jIPg9SsPtkT:bu58it9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, @RaduBerinde, and @sumeerbhola)

___
*[pkg/kv/kvserver/tenantrate/limiter.go, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-MXcDMcq7XB2QdKgl1l4:-MXgfey6-qGT1AyuX1Cn:b-mevdnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3ea561a2d34848013f70f9b92d7d31ce98ea7/pkg/kv/kvserver/tenantrate/limiter.go#L136)):*
It seems that a lot of the commentary rotted. Especially the commentary on the `Merge. The token-bucket functionality was added later. I needed a token bucket and wanted some other properties I already had here and realized that it was a rather minor code change. I'm going to take a stab at fixing up the commentary.

> I can see why generics would improve the situation, but the Merge method would still be clunky -- it is trying to do many things: config changes, unilateral resource acquisition, returning of resources for non-token-bucket.

Ack, yeah, really it should be the structure doing the updating to the resource rather than the other way around. I made https://github.com/cockroachdb/cockroach/pull/63227 to try to clean that up. 

----

For what it's worth, I can see an argument against doing anything generic here. I found it interesting and valuable when hacking around with different policies for queuing. It may be under-motivated. If we get rid of this usage of the abstract pool then there won't be any users and perhaps we should just unexport it. However, I did use it to good effect in this Friday commit to block instead of explode when failing to allocate memory in a long-running changefeed: https://github.com/cockroachdb/cockroach/pull/63050. In that case the resource was a memory monitor which was dynamically growing and shrinking as opposed to some sort of constant.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@ajwerner I'm working on using RateLimiter. How do you recommend handling the config? When admitting a request, I need to use the parameters in the Config to calculate the amount. I don't want to take another lock (or use atomics) just for that..

ajwerner commented:
Huh, good question. I was thinking a RWMutex or an atomic on a pointer to a
config. If you feel that just using the abstract pool makes it more
straightforward and is less convoluted with the refactor, I’m fine with
that. It also, as you noted, makes the floating point requirement a less
invasive backport.

On Wed, Apr 7, 2021 at 9:00 PM RaduBerinde ***@***.***> wrote:

> @ajwerner <https://github.com/ajwerner> I'm working on using RateLimiter.
> How do you recommend handling the config? When admitting a request, I need
> to use the parameters in the Config to calculate the amount. I don't want
> to take another lock (or use atomics) just for that..
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/62783#issuecomment-815371310>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOBBAQJBY6VZCH7M7IZCQDTHT54FANCNFSM42BETM3A>
> .
>

RaduBerinde commented:
I'll tinker more with it. It's unfortunate, it was much cleaner to just use the RateLimiter.

RaduBerinde committed:
quotapool: extract token bucket accounting

This commit extracts the token bucket accounting logic into a
separate, public type. The same code will be used by the tenantrate
limiter.

Release note: None

RaduBerinde commented:
Ok, so I think the main problem was not the use of the abstract pool per se but the duplication of all the token bucket accounting logic. So I extracted that into a separate TokenBucket type and reused it.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> needs a `make bazel-generate`.

> Reviewed 1 of 19 files at r3, 2 of 16 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXldMcSB9GzcjrniG3t:bhiqxpt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, @RaduBerinde, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 19 files at r3, 4 of 16 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62783#-:-MXsMWjtBvMm0ThsejgM:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @nvanbenschoten, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
TFTRs!
